src/vcpkg/binarycaching.cpp

```diff
@@ -2310,137 +1856,393 @@ namespace
             };
 
             MetricsSubmission metrics;
-            for (const auto& cache_provider : state->binary_cache_providers)
+            for (const auto& cache_provider : s.binary_cache_providers)
             {
                 auto it = metric_names.find(cache_provider);
                 if (it != metric_names.end())
                 {
                     metrics.track_define(it->second);
                 }
             }
-
-            get_global_metrics_collector().track_submission(std::move(metrics));
+
+            get_global_metrics_collector().track_submission(std::move(metrics));
+
+            s.nuget_prefix = args.nuget_id_prefix.value_or("");
+            if (!s.nuget_prefix.empty()) s.nuget_prefix.push_back('_');
+            s.use_nuget_cache = args.use_nuget_cache.value_or(false);
+            s.nuget_repo_info = get_nuget_repo_info_from_env(args);
+
+            auto& fs = paths.get_filesystem();
+            auto& tools = paths.get_tool_cache();
+            const auto& buildtrees = paths.buildtrees();
+
+            ret.nuget_prefix = s.nuget_prefix;
+
+            Optional<ZipTool> lazy_zip_tool;
+            auto zip_tool = [&]() -> const ZipTool& {
+                if (!lazy_zip_tool) lazy_zip_tool.emplace(fs, tools, stdout_sink);
+                return *lazy_zip_tool.get();
+            };
+
+            std::shared_ptr<const GcsStorageTool> lazy_gcs_tool;
```

Comment:
I've reworked this section to avoid the lambdas; please take a look.

Reply:
To be clear I was more trying to avoid the repetition of the 'algorithm' rather than lambdas (although I do have a preference to avoid lambdas where possible because debuggers hate them).

I think this one has been addressed with your rework 👍